File: 004.Median of Two Sorted Arrays.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        size = len(nums1) + len(nums2)
        if size == 0:
            return 0
        elif size == 1:
            if len(nums1) == 0:
                return nums2[0]
            else:
                return nums1[0]


        mid_inx = size / 2
        first = second = last = None
        i = j = 0
        while i < len(nums1) and j < len(nums2):
            if nums1[i] > nums2[j]:
                last = nums2[j]
                j += 1
            else:
                last = nums1[i]
                i += 1
 
            if i + j == mid_inx:
                first = last
            elif i + j == mid_inx + 1:
                second = last
                break

        if j < len(nums2) and i == len(nums1):
            while j < len(nums2):
                last = nums2[j]
                j += 1
                if i + j == mid_inx:
                    first = last
                elif i + j == mid_inx + 1:
                    second = last
                    break

        elif i < len(nums1) and j == len(nums2):
            while i < len(nums1):
                last = nums1[i]
                i += 1
                if i + j == mid_inx:
                    first = last
                elif i + j == mid_inx + 1:
                    second = last
                    break

        if size % 2 == 0:
            return float(first + second) / 2
        else:
            return second

File: 004. Median of Two Sorted Arrays.java
/*
    There are two sorted arrays nums1 and nums2 of size m and n respectively.

    Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

    Example 1:
    nums1 = [1, 3]
    nums2 = [2]

    The median is 2.0

    Example 2:
    nums1 = [1, 2]
    nums2 = [3, 4]

    The median is (2 + 3)/2 = 2.5
 */

    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int index1 = 0, index2 = 0;
        int count = (nums1.length + nums2.length) / 2 + 1;
        boolean isOne = (nums1.length + nums2.length) % 2 == 1;
        int smaller = 0;
        int current = 0;
        while (count != 0) {
            if (nums1.length > index1 && nums2.length > index2) {
                if (nums1[index1] < nums2[index2]) {
                    smaller = current;
                    current = nums1[index1];
                    index1++;
                } else {
                    smaller = current;
                    current = nums2[index2];
                    index2++;
                }
            } else if (nums1.length > index1) {
                smaller = current;
                current = nums1[index1];
                index1++;
            } else {
                smaller = current;
                current = nums2[index2];
                index2++;
            }
            count--;
        }

        if (isOne)
            return current/1.0;

        return (smaller + current) / 2.0;
    }
